ruby_programming/git/advanced_git_lesson.md

```diff
@@ -0,0 +1,182 @@
+
+~~~bash
+$ git reset HEAD^
+$ git add lib/knight.rb
+$ git commit -m 'Add knight class'
+$ git add lib/movement.rb
+$ git commit -m 'Add attack module'
+$ git rebase --continue
+~~~
+
+As always however, remember to **check that none of these commits are pushed to shared repositories**.
+
+### Working With Remotes
+
+Thus far you've been working with remote repositories each time you've pushed or pulled from your own Github repository while working on the curriculum's various projects. In this section we're going to cover some slightly more advanced topics, which you might not have yet encountered or had to use. 
+
+## git push --force
+
+Let's say you're no longer working on a project all by yourself, but with someone else. You want to push a branch you've made changes on to a remote repository. Normally Git will only let you push your changes if you've already updated your local branch with the latest commits from this remote. 
+
+If you haven't updated your local branch, and you're attempting to `git push` a commit which would create a conflict on the remote repository, you'll get an error message. This is actually a great thing! This is a safety mechanism to prevent you from overwriting commits created by the people you're working with, which could be disasterous. You get the error because your history is outdated. 
+
+You might perform a brief query and find the command `git push --force`. This command overwrites the remote repository with your own local history. So what would happen we used this while working with others? You could potentially destroy their work! `git push --force` is a **very dangerous command, and should be used with caution when collaborating with others**. Instead, you can your outdated history error by updating your local history using `fetch`, `merge`, and then attempting to `push` again. 
+
+Let's say alternatively that you pushed a commit, and *oops* made a mistake. You want to undo this commit and are once again tempted to just force the push. But wait, remember, this is a **very dangerous command**. If you're ever considering using it always check if it's appropriate and if you can use a safer command instead. If you're collaborating with others and want to *undo* a commit you just made, you can instead use `git revert`! 
+
+~~~bash
+git revert HEAD~1
+git push origin feature-branch
+~~~
+
+Remember when we were working with HEAD, aka the current commit we're viewing, while rebasing? What this would does is it would revert the changes to HEAD by 1 commit! Then we would push our new commit to whichever branch we're working on, which in this example is feature-branch. 
```

Comment:
```suggestion
Remember when we were working with HEAD, aka the current commit we're viewing, while rebasing? What this would do is it would revert the changes to HEAD by 1 commit! Then we would push our new commit to whichever branch we're working on, which in this example is feature-branch. 
```